### ISSUE TITLE
Why not use `TestCoroutineDispatcher`?

### ISSUE THREAD

**Issue by erikhuizinga** [uniflow-kt/uniflow-kt#13]
Use case:

`TestDispatchersRule` uses `Dispatchers.Unconfined`. A user writing tests might use a `TestCoroutineDispatcher` to test their coroutines. If used together with `TestDispatchersRule`, the user now is using two different coroutine dispatchers. This might lead to undetected leaking coroutines or to difficult to debug errors.

This is just a hypothesis. I have no working example that shows that there can(not) be problems. Please let me know what you think that could happen when using both `Dispatchers.Unconfined` and `TestCoroutineDispatcher` in tests.

If there is a particular reason not to use `TestCoroutineDispatcher` or to only use `Dispatchers.Unconfined`, let me know, as I'm unaware of it.

---

AFAIK `TestCoroutineDispatcher` gives more control over the coroutines, but it also introduces a bit of boilerplate:

 - The `Main` dispatcher must be replaced using `Dispatchers.setMain()` before each test.
    - It must also be reset after each test using `Dispatchers.resetMain()`
 - Test functions using coroutines should be wrapped in `runBlockingTest` ([not required, but recommendable](https://github.com/Kotlin/kotlinx.coroutines/tree/1.3.3/kotlinx-coroutines-test#using-testcoroutinescope-and-testcoroutinedispatcher-without-runblockingtest)).
 - After each test `cleanupTestCoroutines()` should be called on the `TestCoroutineDispatcher`.

Similarly, there is `TestCoroutineScope` that behaves similarly to `TestCoroutineDispatcher`.

---

Relevant code:

https://github.com/uniflow-kt/uniflow-kt/blob/6481d4a1e0ab47e1b8927ebc3d252dc2e43ba440/uniflow-core/src/main/kotlin/io/uniflow/core/dispatcher/TestDispatchers.kt#L24-L26

---

Suggestion:
 - Rename `TestDispatchersRule` to `UnconfinedDispatchersRule`, rename `TestDispatchers` to `UnconfinedDispatchers`. This keeps the implementation around as I'm sure it has applications.
 - Use `TestCoroutineDispatcher` in `TestDispatchersRule` and `TestDispatchers`. Also add the necessary before/after steps to be taken. Update the documentation to suggest the use of `TestDispatchersRule` and `runBlockingTest`.

**Comment by costular** [comment]
+1

**Comment by erikhuizinga** [comment]
Also read the secion 'What about Dispatchers.Unconfined?' in [this article](https://medium.com/androiddevelopers/testing-two-consecutive-livedata-emissions-in-coroutines-5680b693cbf8).

**Comment by arnaudgiuliani** [comment]
I would go for the 2nd 👍 

```
Use TestCoroutineDispatcher in TestDispatchersRule and TestDispatchers. Also add the necessary before/after steps to be taken. Update the documentation to suggest the use of TestDispatchersRule and runBlockingTest.
```

**Comment by erikhuizinga** [comment]
I could make a PR, then assign me. Or address this yourself.

**Comment by arnaudgiuliani** [comment]
You can go for a proposal 👍

**Comment by erikhuizinga** [comment]
@costular now implemented in #30.

**Comment by costular** [comment]
@erikhuizinga great! 🎉
